### PR TITLE
IRCv3.3 Message Tags

### PR DESCRIPTION
**Pull request by jwheare**
This specification adds a new capability for general message tags support and
a prefix for expressing client-to-client tags.

This PR replaces #236

### CODE REVIEW

File: message-tags-3.3.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Client-to-client tags are client-initiated tags that servers MUST attach as-is
+to any relevant event relayed to other clients. A client-to-client tag is prefixed
+with a plus sign (`+`) and otherwise conforms to the format specified in
+[IRCv3.2 tags][irctags].
+
+Client-to-client tags MUST be relayed on `PRIVMSG` and `NOTICE` events, and
+MAY be forwarded on other events.
+
+The expected client behaviour of individual client-to-client tags SHOULD be defined
+in separate specifications, in the same way as server-initiated tags.
```
Comment (kylef) [inline message-tags-3.3.md]: I'd love to see some clarification in the spec if tag keys themselves within a message are unique or can be used multiple times.

Whether the following is valid/expected or not:

```
@+icon=https://example.com/favicon.png;+icon=something :url_bot!bot@example.com PRIVMSG #channel :Example.com: A News Story
```

File: message-tags-3.3.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+### Tags
+
+Client-to-client tags are client-initiated tags that servers MUST attach as-is
+to any relevant event relayed to other clients. A client-to-client tag is prefixed
+with a plus sign (`+`) and otherwise conforms to the format specified in
+[IRCv3.2 tags][irctags].
+
+Client-to-client tags MUST be relayed on `PRIVMSG` and `NOTICE` events, and
+MAY be forwarded on other events.
```
Comment (DarthGandalf) [inline message-tags-3.3.md]: Relayed vs forwarded. Let's use a single term for single meaning.

File: core/message-tags-3.3.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+This specification adds the `message-tags` capability. Clients which request
+this capability indicate that they are capable of parsing all tags, regardless
+of support for tags specified in other IRCv3.2 or later specifications.
+
+### Tags
+
+Client-to-client tags are client-initiated tags that servers MUST attach as-is
+to any relevant event relayed to other clients. A client-to-client tag is prefixed
+with a plus sign (`+`) and otherwise conforms to the format specified in
+[IRCv3.2 tags][irctags].
```
Comment (DanielOaks) [inline core/message-tags-3.3.md]: `[IRCv3.2 tags](./message-tags-3.2.html).`

**Comment by DanielOaks** [comment]
Looks good to me, +1 on merging.

**Comment by jwheare** [comment]
* Added new sections for client implementation and moderation considerations.
* Rewrote the motivation section.
* Renamed "client-to-client" to "client-only", less of a mouthful.

**Comment by jwheare** [comment]
Last call for comments.

**Comment by grawity** [comment]
Remind me, how does a _client_ know that it can send tags? Does the new cap work in both directions?

Looks good overall.

**Comment by jwheare** [comment]
> Remind me, how does a client know that it can send tags? Does the new cap work in both directions?

Yes. I'll clarify.

File: core/message-tags-3.3.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+The expected client behaviour of individual client-only tags SHOULD be defined
+in separate specifications, in the same way as server-initiated tags.
+
+The updated BNF for keys is as follows:
+
+    <key>           ::= [ '+' ] [ <vendor> '/' ] <sequence of letters, digits, hyphens (`-`)>
+
+Individual tag keys MUST only be used a maximum of once per message. Clients
+receiving messages with more than one occurrence of a tag key SHOULD discard all
+but the final occurrence.
```
Comment (SaberUK) [inline core/message-tags-3.3.md]: Perhaps we should add a reminder around here that people should use vendor prefixes for proprietary features.

Comment (jwheare) [inline core/message-tags-3.3.md]: Done.

**Comment by SaberUK** [comment]
LGTM. :shipit:

**Comment by jwheare** [comment]
Squashed

**Comment by dequis** [comment]
LGTM

**Comment by attilamolnar** [comment]
LGTM

**Comment by SaberUK** [comment]
LGTM :shipit:

**Pull request merged**